File: Aulas/Aula5/Composicao_exemplo.java
// Classe Motor
class Motor {
    private String tipo;
    private int potencia;

    public Motor(String tipo, int potencia) {
        this.tipo = tipo;
        this.potencia = potencia;
    }

    public void ligar() {
        System.out.println("O motor está ligado.");
    }

    public void desligar() {
        System.out.println("O motor está desligado.");
    }
}

// Classe Pneu
class Pneu {
    private String marca;
    private int tamanho;

    public Pneu(String marca, int tamanho) {
        this.marca = marca;
        this.tamanho = tamanho;
    }

    public void inflar() {
        System.out.println("O pneu está inflado.");
    }

    public void desinflar() {
        System.out.println("O pneu está desinflado.");
    }
}

// Classe Carro
class Carro {
    private String marca;
    private String modelo;
    private Motor motor;  // Composição: Carro tem um Motor
    private Pneu[] pneus; // Composição: Carro tem 4 Pneus

    public Carro(String marca, String modelo) {
        this.marca = marca;
        this.modelo = modelo;
        this.motor = new Motor("Gasolina", 150);
        this.pneus = new Pneu[4];
        for (int i = 0; i < 4; i++) {
            pneus[i] = new Pneu("Pirelli", 18);
        }
    }

    public void ligar() {
        motor.ligar();
        System.out.println("O carro está pronto para rodar.");
    }

    public void desligar() {
        motor.desligar();
        System.out.println("O carro foi desligado.");
    }

    public void trocarPneu(int indice, Pneu novoPneu) {
        if (indice >= 0 && indice < 4) {
            pneus[indice] = novoPneu;
            System.out.println("Pneu " + (indice + 1) + " trocado.");
        } else {
            System.out.println("Índice de pneu inválido.");
        }
    }
}

// Cliente
public class Composicao_exemplo{
    public static void main(String[] args) {
        Carro carro = new Carro("Toyota", "Corolla");
        carro.ligar();

        // Trocando o pneu da posição 2
        Pneu novoPneu = new Pneu("Michelin", 17);
        carro.trocarPneu(2, novoPneu);

        carro.desligar();
    }
}

File: Aulas/Aula5/Composicao_exemplo.go
package main

import "fmt"

// Struct Motor
type Motor struct {
    tipo    string
    potencia int
}

func (m Motor) Ligar() {
    fmt.Println("O motor está ligado.")
}

func (m Motor) Desligar() {
    fmt.Println("O motor está desligado.")
}

// Struct Pneu
type Pneu struct {
    marca  string
    tamanho int
}

func (p Pneu) Inflar() {
    fmt.Println("O pneu está inflado.")
}

func (p Pneu) Desinflar() {
    fmt.Println("O pneu está desinflado.")
}

// Struct Carro
type Carro struct {
    marca  string
    modelo string
    motor  Motor
    pneus  [4]Pneu
}

func NewCarro(marca, modelo string) Carro {
    motor := Motor{"Gasolina", 150}
    pneus := [4]Pneu{}
    for i := 0; i < 4; i++ {
        pneus[i] = Pneu{"Pirelli", 18}
    }
    return Carro{marca, modelo, motor, pneus}
}

func (c Carro) Ligar() {
    c.motor.Ligar()
    fmt.Println("O carro está pronto para rodar.")
}

func (c Carro) Desligar() {
    c.motor.Desligar()
    fmt.Println("O carro foi desligado.")
}

func (c *Carro) TrocarPneu(indice int, novoPneu Pneu) {
    if indice >= 0 && indice < 4 {
        c.pneus[indice] = novoPneu
        fmt.Printf("Pneu %d trocado.\n", indice+1)
    } else {
        fmt.Println("Índice de pneu inválido.")
    }
}

// Cliente
func main() {
    carro := NewCarro("Toyota", "Corolla")
    carro.Ligar()

    // Trocando o pneu da posição 2
    novoPneu := Pneu{"Michelin", 17}
    carro.TrocarPneu(2, novoPneu)

    carro.Desligar()
}
